File: src/telemetry.py
"""Structured telemetry utilities for the shell assistant."""

from __future__ import annotations

import json
import os
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path
from typing import Any, Dict, Optional


@dataclass
class TelemetryEmitter:
    """Writes structured telemetry events as line-delimited JSON."""

    path: Path
    enabled: bool

    @classmethod
    def initialize(cls, file_path: Optional[str]) -> "TelemetryEmitter":
        if file_path == "":
            return cls(path=Path("."), enabled=False)

        candidate = file_path or os.getenv("AGENT_TELEMETRY_FILE") or "logs/telemetry.jsonl"
        path = Path(candidate).expanduser()
        if not path.is_absolute():
            path = Path.cwd() / path
        path.parent.mkdir(parents=True, exist_ok=True)
        return cls(path=path, enabled=True)

    def emit(self, payload: Dict[str, Any]) -> None:
        if not self.enabled:
            return
        record = dict(payload)
        record.setdefault("timestamp", datetime.utcnow().isoformat(timespec="seconds") + "Z")
        try:
            with self.path.open("a", encoding="utf-8") as handle:
                json.dump(record, handle, ensure_ascii=True)
                handle.write("\n")
        except OSError:
            # Telemetry is best-effort; ignore write failures silently for now.
            return

    def emit_execution(
        self,
        goal: str,
        session_id: Optional[str],
        planner_info: Dict[str, Any],
        command_event: Dict[str, Any],
    ) -> None:
        if not self.enabled:
            return
        event = {
            "event": "execution",
            "goal": goal,
            "session_id": session_id,
            "planner_info": planner_info,
            **command_event,
        }
        self.emit(event)

    def emit_plan_created(
        self,
        goal: str,
        session_id: Optional[str],
        planner_info: Dict[str, Any],
        plan: Dict[str, Any],
    ) -> None:
        if not self.enabled:
            return
        event = {
            "event": "plan_created",
            "goal": goal,
            "session_id": session_id,
            "planner_info": planner_info,
            "plan": plan,
        }
        self.emit(event)

    def emit_plan_update(
        self,
        goal: str,
        session_id: Optional[str],
        planner_info: Dict[str, Any],
        plan: Dict[str, Any],
        step_id: Optional[str],
        status: Optional[str],
    ) -> None:
        if not self.enabled:
            return
        event = {
            "event": "plan_updated",
            "goal": goal,
            "session_id": session_id,
            "planner_info": planner_info,
            "plan": plan,
            "step_id": step_id,
            "status": status,
        }
        self.emit(event)

File: src/lib/telemetry.ts
/**
 * Structured telemetry utilities for the shell assistant.
 */

import fs from "fs";
import path from "path";
import type { Logger } from "./logger.js";

export interface TelemetryEvent {
  event: string;
  timestamp?: string;
  [key: string]: any;
}

export interface TelemetryEmitter {
  path: string;
  enabled: boolean;

  emit(payload: Record<string, any>): void;
  emitExecution(
    goal: string,
    sessionId: string | null,
    plannerInfo: Record<string, any>,
    commandEvent: Record<string, any>
  ): void;
  emitPlanCreated(
    goal: string,
    sessionId: string | null,
    plannerInfo: Record<string, any>,
    plan: Record<string, any>
  ): void;
  emitPlanUpdate(
    goal: string,
    sessionId: string | null,
    plannerInfo: Record<string, any>,
    plan: Record<string, any>,
    stepId: string | null,
    status: string | null
  ): void;
}

/**
 * Redacts common secret patterns from a string before it is persisted to the
 * telemetry log. Targets inline credentials that may appear in command strings
 * or error messages: KEY=value pairs, Authorization headers, -p<password>.
 */
function scrubSecrets(text: string): string {
  return text
    // KEY=value or KEY: value where KEY name looks like a credential
    .replace(
      /\b(password|passwd|token|api[_-]?key|secret|credential|private[_-]?key)s?\s*[=:]\s*\S+/gi,
      (_, key: string) => `${key}=[REDACTED]`
    )
    // Authorization header (Bearer / Basic / etc.)
    .replace(/\bAuthorization:\s*\S+(\s+\S+)?/gi, "Authorization: [REDACTED]")
    // MySQL-style -p<password> flag
    .replace(/(\s)-p\S+/g, "$1-p[REDACTED]");
}

export class TelemetryEmitterImpl implements TelemetryEmitter {
  path: string;
  enabled: boolean;
  private _logger: Logger | null = null;
  /** Tracks the last error message to avoid spamming the log on every emit. */
  private _lastErrorMsg: string | null = null;

  constructor(filePath: string, enabled: boolean = true) {
    this.path = filePath;
    this.enabled = enabled;
  }

  setLogger(logger: Logger): void {
    this._logger = logger;
  }

  static initialize(filePath?: string): TelemetryEmitterImpl {
    if (filePath === "") {
      return new TelemetryEmitterImpl(".", false);
    }

    let candidate = filePath || process.env.AGENT_TELEMETRY_FILE || "logs/telemetry.jsonl";
    
    // Ensure path is absolute
    const absolutePath = path.isAbsolute(candidate) ? candidate : path.join(process.cwd(), candidate);
    
    // Create parent directories
    const dir = path.dirname(absolutePath);
    fs.mkdirSync(dir, { recursive: true });
    
    return new TelemetryEmitterImpl(absolutePath, true);
  }

  emit(payload: Record<string, any>): void {
    if (!this.enabled) {
      return;
    }
    
    const record = { ...payload };
    if (record.timestamp == null) record.timestamp = new Date().toISOString();
    
    // Scrub secrets from all top-level string fields before persisting.
    const clean: Record<string, any> = {};
    for (const [k, v] of Object.entries(record)) {
      clean[k] = typeof v === "string" ? scrubSecrets(v) : v;
    }
    const line = JSON.stringify(clean) + "\n";
    try {
      // Open in append mode and capture pre-write size so we can roll back a
      // partial/corrupt write (e.g. mid-crash) by truncating to prevSize.
      const fd = fs.openSync(this.path, "a");
      try {
        const prevSize = fs.fstatSync(fd).size;
        try {
          fs.writeSync(fd, line);
        } catch (writeErr) {
          try { fs.ftruncateSync(fd, prevSize); } catch { /* best-effort rollback */ }
          throw writeErr;
        }
      } finally {
        fs.closeSync(fd);
      }
      this._lastErrorMsg = null; // clear dedup key on success
    } catch (error) {
      const msg = error instanceof Error ? error.message : String(error);
      if (msg !== this._lastErrorMsg) {
        this._lastErrorMsg = msg;
        this._logger?.warning(`Telemetry write failed (${this.path}): ${msg}`);
      }
    }
  }

  emitExecution(
    goal: string,
    sessionId: string | null,
    plannerInfo: Record<string, any>,
    commandEvent: Record<string, any>
  ): void {
    if (!this.enabled) {
      return;
    }
    
    const event = {
      event: "execution",
      goal,
      session_id: sessionId,
      planner_info: plannerInfo,
      ...commandEvent,
    };
    
    this.emit(event);
  }

  emitPlanCreated(
    goal: string,
    sessionId: string | null,
    plannerInfo: Record<string, any>,
    plan: Record<string, any>
  ): void {
    if (!this.enabled) {
      return;
    }
    
    const event = {
      event: "plan_created",
      goal,
      session_id: sessionId,
      planner_info: plannerInfo,
      plan,
    };
    
    this.emit(event);
  }

  emitPlanUpdate(
    goal: string,
    sessionId: string | null,
    plannerInfo: Record<string, any>,
    plan: Record<string, any>,
    stepId: string | null,
    status: string | null
  ): void {
    if (!this.enabled) {
      return;
    }
    
    const event = {
      event: "plan_updated",
      goal,
      session_id: sessionId,
      planner_info: plannerInfo,
      plan,
      step_id: stepId,
      status,
    };
    
    this.emit(event);
  }
}

